houston/index.js

```diff
@@ -0,0 +1,163 @@
+import view from 'koa-views'
+
+import * as helpers from '~/lib/helpers'
+import * as passport from './passport'
+import atc from './service/atc'
+import config from '~/lib/config'
+import controllers from './controller'
+import db from '~/lib/database'
+import log from '~/lib/log'
+import Mistake from '~/lib/mistake'
+
+const app = new Koa()
+
+// Setup App configuration
+app.name = 'Houston'
+app.env = config.env
+
+// Socket installation
+const server = http.createServer(app.callback())
+atc.connect(server)
+
+// App logging
+app.use(async (ctx, next) => {
+  const start = new Date()
+  await next()
+  const end = new Date()
+  log.verbose(`${ctx.method} ${ctx.status} ${ctx.url} => ${end - start}ms`)
+})
+
+// Static 'public' folder serving
+app.use(koaStatic(path.join(__dirname, 'public')))
+
+// Setup server rendering
+app.use(convert(view(path.join(__dirname, 'views'), {
+  extension: 'jade',
+  cache: (app.env === 'production')
+})))
+
+app.use(async (ctx, next) => {
+  ctx.render = co.wrap(ctx.render)
+  ctx.Mistake = Mistake
+
+  ctx.state.basedir = path.normalize(`${__dirname}/views`)
+  ctx.state.config = config
+  ctx.state.helper = helpers
+
+  ctx.state.title = 'Houston'
+  await next()
+})
+
+// Error pages
+// eslint-disable-next-line consistent-return
+app.use(async (ctx, next) => {
+  try {
+    await next()
+  } catch (error) {
+    ctx.app.emit('error', error, ctx)
+
+    const htmlRespond = (ctx.accepts(['json', 'html']) === 'html')
+
+    const pkg = {
+      status: error.status || 500
+    }
+
+    if (app.env === 'development') {
+      pkg.detail = error.stack
+    }
+
+    if (error.expose) {
+      pkg.title = error.message || 'Houston has encountered an error'
+    } else {
+      pkg.title = 'Houston has encountered an error'
+    }
+
+    ctx.status = pkg.status
+    if (htmlRespond) {
+      return ctx.render('error', { error: pkg })
+    } else {
+      ctx.body = { errors: [pkg] }
+      return null
+    }
+  }
+})
+
+// Start Passport
+app.use(parser())
+app.keys = [config.server.secret]
+app.use(convert(session(app)))
+
+passport.setup(app)
+
+app.use(passport.router.routes(), passport.router.allowedMethods())
+
+// Load Houston core files
+app.use(controllers.routes(), controllers.allowedMethods())
+
+// 404 page
+app.use((ctx) => {
+  ctx.status = 404
+
+  if (ctx.accepts(['json', 'html']) === 'html') {
+    return ctx.render('error', { error: {
+      status: 404,
+      title: 'Page not found',
+      detail: ''
+    }})
+  } else {
+    ctx.body = { errors: [{
+      status: 404,
+      title: 'Page Not Found',
+      detail: 'The page you are looking found can not be found'
+    }]}
+    return null
+  }
+})
+
+// Error logging
+app.on('error', async (error, ctx, next) => {
+  if (app.env === 'test') return
+
+  if (/4.*/.test(error.status)) {
+    log.verbose(`${ctx.method} ${ctx.status} ${ctx.url} |> ${error.message}`)
+  } else {
+    log.error(error)
+  }
+
+  try {
+    await next()
```

Comment:
What exactly would be the next error handler, and why exactly are you waiting on it to return a promise?